New in town challenges

Catalin-Gruia opened the pull request:
Pull request for "New in town" challenges.

Catalin-Gruia committed:
Working implementation of Capitalize sentence

Catalin-Gruia committed:
Working implementation of Get odd numbers

Catalin-Gruia committed:
Working implementation of Contains range?

Catalin-Gruia committed:
Working implementation of Add up to

Catalin-Gruia committed:
Working implementation of Linear search

Catalin-Gruia committed:
Working implementation of Count down

Catalin-Gruia committed:
Working implementation of Count up and down

Catalin-Gruia committed:
Working implementation of Format train route

Catalin-Gruia committed:
Working implementation of Generate all pairs

Catalin-Gruia committed:
Working implementation of Count unique values

Catalin-Gruia committed:
Working implementation of Print number

Catalin-Gruia committed:
Working implementation of Print number with steps

Catalin-Gruia committed:
Working implementation of Power

Catalin-Gruia committed:
Working implementation of Factorial

Catalin-Gruia committed:
Working implementation of Product

Catalin-Gruia committed:
Working implementation of Capitalize first

Catalin-Gruia committed:
Working implementation of Decapitalize const

Catalin-Gruia committed:
Working implementation of Longest word

Catalin-Gruia committed:
Working implementation of Is palindrome

First Working implementation of Is palindrome

radu-marin-sv commented on src/test/kotlin/com/igorwojda/integer/addupto/challenge.kt:
```diff
@@ -4,7 +4,8 @@ import org.amshove.kluent.shouldBeEqualTo
 private fun addUpTo(n: Int): Int {
-    TODO("not implemented")
+    // Gauss formula for consecutive numbers 1 + 2 + 3 + 4 + … + n == n x ( n + 1 ) : 2
+    return n*(n+1)/2
```
nitpick: this is indeed the most optimal solution. I'd also suggest looking into aggregate operations: https://kotlinlang.org/docs/collection-aggregate.html

radu-marin-sv commented on src/test/kotlin/com/igorwojda/integer/factorial/challenge.kt:
```diff
@@ -4,7 +4,10 @@ import org.amshove.kluent.shouldBeEqualTo
 
 private fun factorial(n: Int): Int {
-    TODO("not implemented")
+    var total = 1
```
Re-write this:
1. Using a folding operation
2. Using a recursive algorithm
3. If you're curious, look at tail recursive functions (https://kotlinlang.org/docs/reference/functions.html#tail-recursive-functions). See if you can write factorial as a tailrec function.
Please compare the running time of the solutions.

radu-marin-sv commented on src/test/kotlin/com/igorwojda/integer/power/challenge.kt:
```diff
@@ -2,9 +2,10 @@ package com.igorwojda.integer.power
 
 private fun power(base: Int, exponent: Int): Int {
-    TODO("not implemented")
+    return base.toDouble().pow(exponent.toDouble()).toInt()
```
Challenge: reimplement this by using the following concepts:
1. List initializer: https://kotlinlang.org/docs/reference/constructing-collections.html#initializer-functions-for-lists
2. Reduce operation: https://kotlinlang.org/docs/reference/collection-aggregate.html#fold-and-reduce
3. An operator: https://kotlinlang.org/docs/reference/operator-overloading.html
4. A method reference (to the above operator): https://khan.github.io/kotlin-for-python-developers/#function-references

It should be a one-liner ;)

radu-marin-sv commented on src/test/kotlin/com/igorwojda/list/capitalizeFirst/challenge.kt:
```diff
@@ -4,7 +4,9 @@ import org.amshove.kluent.shouldBeEqualTo
 
 private fun capitalizeFirst(list: List<String>): List<String> {
-    TODO("not implemented")
+    return if (list.isEmpty())
```
Please look at collection transformation operations: https://kotlinlang.org/docs/collection-transformations.html

radu-marin-sv commented on src/test/kotlin/com/igorwojda/list/countuniquevalues/challenge.kt:
```diff
@@ -4,7 +4,7 @@ import org.amshove.kluent.shouldBeEqualTo
 
 private fun countUniqueValues(list: List<Int>): Int {
-    TODO("not implemented")
+    return list.distinct().size
```
challenge: convert the List to a different type of collection that yields the unique items in the list. Can you think which one it is?

radu-marin-sv commented on src/test/kotlin/com/igorwojda/list/formattrainroute/challenge.kt:
```diff
@@ -4,7 +4,22 @@ import org.amshove.kluent.shouldBeEqualTo
 
 private fun formatTrainRoute(stations: List<String>): String {
-    TODO("not implemented")
+    if (stations.size > 1) {
```
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html is a very powerful tool

radu-marin-sv commented on src/test/kotlin/com/igorwojda/list/product/challenge.kt:
```diff
@@ -4,7 +4,11 @@ import org.amshove.kluent.shouldBeEqualTo
 
 private fun product(list: List<Int>): Int {
-    TODO("not implemented")
+    var product = 1
```
Challenge: reimplement this by using the following concepts:
1. Reduce operation: https://kotlinlang.org/docs/reference/collection-aggregate.html#fold-and-reduce
2. An operator: https://kotlinlang.org/docs/reference/operator-overloading.html
3. A method reference (to the above operator): https://khan.github.io/kotlin-for-python-developers/#function-references
It should be a single chain of functional calls.

radu-marin-sv commented on src/test/kotlin/com/igorwojda/list/search/linearsearch/challenge.kt:
```diff
@@ -4,7 +4,10 @@ import org.amshove.kluent.shouldBeEqualTo
 
 private fun getIndex(list: List<String>, str: String): Int {
-    TODO("not implemented")
+    for (item in list.indices)
```
nitpick: there's also a forEachIndexed on collections

radu-marin-sv reviewed (changes requested):
Before addressing the comments, please check out https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods, especially for formatting and indentation

Catalin-Gruia committed:
Updater implementation after Core-Review

Refined implementation after Core-Review

Catalin-Gruia committed:
Refined code style

Refined code style after reading Kotlin naming conventions

Catalin-Gruia committed:
Refined implementation

Better indentation and added implementation using aggregate operations

Catalin-Gruia committed:
Refined formating

Added curly bracket for if condition

Catalin-Gruia committed:
Refined implementation using join to string internal methods

Refined implementation using join to string internal methods

Catalin-Gruia committed:
Refined formating

Removed spaces from inside of range operator
Kotlin conventions -> "don't put spaces around the "range to" operator (0..i)."

Catalin-Gruia committed:
Changed implementation from List to Map

Changed implementation from List to Map so that it contains unique items

Catalin-Gruia committed:
Refined implementation using all required concepts

Refined implementation using all required concepts from Code Review

Catalin-Gruia committed:
Added factorial using folding operations and using recursive algorithm

1. Re-write this: Using a folding operation
2. Re-write this: Using a recursive algorithm

Catalin-Gruia committed:
Reimplement this by using the following concepts

Reimplement this by using the following concepts
1. Reduce operation
2. An operator
3. A method reference

Catalin-Gruia committed:
Reimplemented using collection transformation operations

Reimplemented using collection transformation operations

Catalin-Gruia committed:
Refined implementation from get(i) to .[i]

Refined implementation from get(i) to .[i]

Catalin-Gruia committed:
Revert "Working implementation of Is palindrome"

This reverts commit 75eaf1ebdb6c9f4459b3c95e28474db5291be103.

radu-marin-sv commented on src/test/kotlin/com/igorwojda/integer/power/challenge.kt:
```diff
@@ -4,7 +4,12 @@ import org.amshove.kluent.shouldBeEqualTo
+    //    return List(exponent) { base }.reduce { product, element -> product * element }
+
+    // But since you asked to have all the required concepts, please check the implementation below :)
+    operator fun List<Int>.invoke() = reduce { it, product -> it * product }
```
Why is this line required?

Catalin-Gruia replied:
Yes, you are right. It is not required, I can implement this challenge with only the following line:

` return List(exponent) { base }.reduce(Int::times)`

radu-marin-sv commented on src/test/kotlin/com/igorwojda/list/product/challenge.kt:
```diff
@@ -4,7 +4,12 @@ import org.amshove.kluent.shouldBeEqualTo
+    //    return list.reduce { it, product -> product * it }
+
+    // But since you asked to have all the required concepts, please check the implementation below :)
+    operator fun List<Int>.invoke() = reduce { it, product -> it * product }
```
Why is this line required?

Catalin-Gruia replied:
Yes, you are right. It is not required, I can implement this challenge with only the following line:

`return list.reduce(Int::times)`

radu-marin-sv commented:
Please use "Squash and merge"
